proposals/tf-operator-design-v1alpha2.md

```diff
@@ -0,0 +1,435 @@
+
+## Goals
+
+- Define the structure of API `v1alpha2`.
+    + Cover most of the refactoring requests we have discussed.
+    + Simplify the API definition.
+- Define an `event-driven` mechanism for TFJob life-cycle management.
+    + And use `reconciler` mechanism as a double check.
+- Clarify the `error handing` logic.
+- Provide a `test` mechanism to verify the design and implementation.
+
+## Non-Goals
+
+- **Notes:** As we make a big change in API v1alpha2, compatibility with v1alpha1 is **NOT** be taken into consideration in this proposal.
+
+## UI or API
+
+The `TFJob` API v1alpha2 object will have the following structure:
+
+**TFJob**:
+```go
+// TFJob represents the configuration of signal TFJob
+type TFJob struct {
+    metav1.TypeMeta `json:",inline"`
+
+    // Standard object's metadata.
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Specification of the desired behavior of the TFJob.
+    Spec TFJobSpec `json:"spec,omitempty"`
+
+    // Most recently observed status of the TFJob.
+    // This data may not be up to date.
+    // Populated by the system.
+    // Read-only.
+    Status TFJobStatus `json:"status,omitempty"`
+}
+```
+
+**TFJobSpec**:
+```go
+// TFJobSpec is a desired state description of the TFJob.
+type TFJobSpec struct {
+    // TFReplicaSpecs is map of TFReplicaType and TFReplicaSpec
+    // specifies the TF replicas to run.
+    // For example,
+    //   {
+    //     "PS": TFReplicaSpec,
+    //     "Worker": TFReplicaSpec,
+    //   }
+    TFReplicaSpecs map[TFReplicaType]*TFReplicaSpec `json:"tfReplicaSpecs"`
+}
+```
+
+**TFReplicaSpec**:
+```go
+// TFReplicaSpec is a description of the TFReplica
+type TFReplicaSpec struct {
+    // Replicas is the desired number of replicas of the given template.
+    // If unspecified, defaults to 1.
+    Replicas *int32 `json:"replicas,omitempty"`
+
+    // Template is the object that describes the pod that
+    // will be created for this TFReplica.
+    // We use RestartPolicy in PodTemplateSpec
+    // to describe how the containers within the pod should be restarted.
+    // Please set this restart policy carefully according to your code.
+    Template *v1.PodTemplateSpec `json:"template,omitempty"`
+}
+```
+
+**TFReplicaType**:
+```go
+// TFReplicaType is the type for TFReplica.
+type TFReplicaType string
+
+const (
+    // TFReplicaTypePS is the type for parameter servers of distributed TensorFlow.
+    TFReplicaTypePS TFReplicaType = "PS"
+
+    // TFReplicaTypeWorker is the type for workers of distributed TensorFlow.
+    TFReplicaTypeWorker TFReplicaType = "Worker"
+
+    // TFReplicaTypeLocal is the type for local TensorFlow.
+    TFReplicaTypeLocal TFReplicaType = "Local"
+)
+```
+
+**TFJobStatus**:
+```go
+// TFJobStatus represents the current observed state of the TFJob.
+type TFJobStatus struct {
+    // Phase is the recently observed lifecycle phase of the TFJob.
+    Phase TFJobPhase `json:"phase"`
```

Comment:
Do we need Phase if we have conditions? I thought @gaocegege or someone else suggested that K8s was moving away from using Phase and using conditions instead?

Reply:
Phase is required but we should use condition to determine the the state of the TFJob.

Reply:
@gaocegege Required by K8s?

Reply:
@jlewi By convention, I think

Reply:
For reference: kubeflow/tf-operator#223 as quoted in their from the docs

> Some resources in the v1 API contain fields called phase, and associated message, reason, and other status fields. The pattern of using phase is deprecated. 

My interpretation of this is that not all V1 resources contain phase. So I don't think we should include phase.

I think the information in phase overlaps with conditions. So by providing phase and conditions are API isn't as clean as it could be because users will need to figure out which one to look at.